In [1]:
import numpy as np
import pandas as pd
import matplotlib as plt
import matplotlib.pyplot as plt
from matplotlib import pyplot
from numpy import std
import seaborn as sns
from scipy import stats

In [2]:
df = pd.read_csv('philly_restaurants_categories.csv')
df

business_id  category
0    MUTTqe8uqyMdBl186RmNeA  Japanese
1    QdN72BWoyFypdGJhhI5r7g   Italian
2    WYYdQDjx-DsCanlP0DpImQ  Japanese
3    O1oZpbZNDMH_gz8DhsZCdA   Burgers
4    dChRGpit9fM_kZK5pafNyA   Burgers
..                      ...       ...
946  oZzN706lKoL4faaTK739xA   Italian
947  p8FRvYD-vYAKzAiZgbJdbg  Japanese
948  bmJgvpuf2GBXd62ELK0Q2w   Italian
949  gPr1io7ks0Eo3FDsnDTYfg   Italian
950  wVxXRFf10zTTAs11nr4xeA   Italian

[951 rows x 2 columns]

In [3]:
df1 = pd.read_csv('yelp_academic_dataset_reviews.csv')
df1

review_id                 user_id  \
0       BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q   
1       AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ   
2       JrIxlS1TzJ-iCu79ul40cQ  eUta8W_HdHMXPzLBBZhL1A   
3       8JFGBuHMoiNDyfcxuWNtrA  smOvOajNG0lS4Pq7d8g4JQ   
4       oyaMhzBSwfGgemSGuZCdwQ  Dd1jQj7S-BFGqRbApFzCFw   
...                        ...                     ...   
967547  tWCi4N2qCil876BI2NDAZg  5d32aJe0a3_fSl3LIVYkWQ   
967548  nLjbVsETpqO17RbFcqskkA  am7-gkH_PDz598oTdYSD6A   
967549  99ylx-qPUSseITqBye2MpA  -AkziDwQ8hv2COTDBBUpig   
967550  KlHxcAifUF5zDKpJCBrRsw  7ziWZULyiZv2TesYNMFf4g   
967551  cACxcUY_AIsQKkpDRXuqnw  MCzlzlOw7IGbRAKVjJBPtg   

                   business_id  stars  useful  funny  cool  \
0       7ATYjTIgM3jUlt4UM3IypQ      5       1      0     1   
1       kxX2SOes4o-D3ZQBkiMRfA      5       1      0     1   
2       04UD14gamNjLY0IDYVhHJg      1       1      2     1   
3       RZtGWDLCAtuipwaZ-UfjmQ      4       0      0     0   
4       YtSqYv1Q_pOltsVPSx54SA      5       0      0     0   
...                        ...    ...     ...    ...   ...   
967547  EmrOCT7fArKI0JI_XF6aaw      1       5      2     0   
967548  3gVSrS4kffGGZT8oXHsIcw      3       2      0     2   
967549  aunmz06iWvo3bd6MMHEbqg      3       0      0     0   
967550  qQO7ErS_RAN4Vs1uX0L55Q      4       1      0     0   
967551  fcGexL5VH5G2Xw0tRj9uOQ      3       1      1     0   

                                                     text                 date  
0       I've taken a lot of spin classes over the year...  2012-01-03 15:28:18  
1       Wow!  Yummy, different,  delicious.   Our favo...  2015-01-04 00:01:03  
2       I am a long term frequent customer of this est...  2015-09-23 23:10:31  
3       Good food--loved the gnocchi with marinara\nth...  2009-10-14 19:57:14  
4       Tremendous service (Big shout out to Douglas) ...  2013-06-24 11:21:25  
...                                                   ...                  ...  
967547  Temple University blatantly disregards freedom...  2021-08-28 18:46:05  
967548  *Later Yelp* I've only been here once, but I l...  2014-11-03 14:45:46  
967549  Philly has become a dangerous place with the m...  2022-01-18 03:48:44  
967550  ice cream! ice cream sodas, sundaes!! \n\nwant...  2012-10-21 04:08:40  
967551  This is a good pizza option - they deliver thr...  2018-03-13 13:54:48  

[967552 rows x 9 columns]

In [4]:
merged_df = df[['business_id', 'category']].merge(df1[['business_id', 'text']], on='business_id', how='left')

In [5]:
merged_df

business_id  category  \
0       MUTTqe8uqyMdBl186RmNeA  Japanese   
1       MUTTqe8uqyMdBl186RmNeA  Japanese   
2       MUTTqe8uqyMdBl186RmNeA  Japanese   
3       MUTTqe8uqyMdBl186RmNeA  Japanese   
4       MUTTqe8uqyMdBl186RmNeA  Japanese   
...                        ...       ...   
155675  wVxXRFf10zTTAs11nr4xeA   Italian   
155676  wVxXRFf10zTTAs11nr4xeA   Italian   
155677  wVxXRFf10zTTAs11nr4xeA   Italian   
155678  wVxXRFf10zTTAs11nr4xeA   Italian   
155679  wVxXRFf10zTTAs11nr4xeA   Italian   

                                                     text  
0       Stopped in to check out this new spot around t...  
1       I live in the neighborhood and used to order a...  
2       We came here tonight just for a date night. We...  
3       Wow! What a great dining adventure! Huge rolls...  
4       Wonderfully fresh sushi, amazing lobster fried...  
...                                                   ...  
155675  I was online and couldn't decide what to order...  
155676  My partner raved about Primo Hoagies so I was ...  
155677  Well built hoagie, nice and full...bread, meat...  
155678  My boyfriend (vegetarian) and I ordered from P...  
155679  For a place that only makes hoagies how do u f...  

[155680 rows x 3 columns]

In [6]:
grouped_df = merged_df.groupby('business_id')['text'].agg(' '.join).reset_index()
merged_df = grouped_df.merge(merged_df[['business_id', 'category']].drop_duplicates(), on='business_id', how='left')


In [7]:
X= merged_df['text'].tolist()
y= merged_df['category'].tolist()

In [8]:
y

['Japanese',
 'Italian',
 'Japanese',
 'Italian',
 'Italian',
 'Burgers',
 'Burgers',
 'Japanese',
 'Italian',
 'Italian',
 'Burgers',
 'Italian',
 'Burgers',
 'Burgers',
 'Italian',
 'Italian',
 'Italian',
 'Italian',
 'Italian',
 'Italian',
 'Japanese',
 'Italian',
 'Italian',
 'Italian',
 'Japanese',
 'Japanese',
 'Japanese',
 'Japanese',
 'Italian',
 'Burgers',
 'Italian',
 'Italian',
 'Japanese',
 'Italian',
 'Italian',
 'Italian',
 'Italian',
 'Japanese',
 'Italian',
 'Burgers',
 'Italian',
 'Italian',
 'Italian',
 'Burgers',
 'Burgers',
 'Italian',
 'Japanese',
 'Italian',
 'Burgers',
 'Italian',
 'Burgers',
 'Burgers',
 'Italian',
 'Burgers',
 'Italian',
 'Italian',
 'Italian',
 'Burgers',
 'Italian',
 'Burgers',
 'Burgers',
 'Burgers',
 'Burgers',
 'Japanese',
 'Italian',
 'Italian',
 'Italian',
 'Italian',
 'Burgers',
 'Burgers',
 'Japanese',
 'Burgers',
 'Italian',
 'Burgers',
 'Italian',
 'Italian',
 'Burgers',
 'Japanese',
 'Burgers',
 'Italian',
 'Burgers',
 'Italian',
 '

                                                -Step 1-

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix


metrics_avg_logreg = np.zeros(4)
metrics_avg_svm = np.zeros(4)
metrics_avg_kNN = np.zeros(4)
avg_cm_logreg = None
avg_cm_svm = None
avg_cm_kNN= None
le = LabelEncoder()
y = le.fit_transform(y)
#Japanese:2 , Italian:1 , Burgers:0
kfold = KFold(n_splits=5, shuffle=True, random_state=42)

fold_count = 0
# split the data into 5 folds
for train_index, test_index in kfold.split(X):
    fold_count += 1
    X_train, X_test = [X[i] for i in train_index], [X[i] for i in test_index]
    y_train, y_test = [y[i] for i in train_index], [y[i] for i in test_index]
    #tf-idf vectorizer
    vectorizer = TfidfVectorizer(stop_words='english')
    X_train_vectorized = vectorizer.fit_transform(X_train)
    X_test_vectorized = vectorizer.transform(X_test)
 
    #logistic Regression
    logreg =LogisticRegression(random_state=0, solver='lbfgs', multi_class='auto', n_jobs=-1)
    logreg.fit(X_train_vectorized, y_train)
    logreg_pred = logreg.predict(X_test_vectorized)
    logreg_precision, logreg_recall, logreg_fscore, logreg_support = precision_recall_fscore_support(y_test, logreg_pred, average='macro')
    logreg_acc = logreg.score(X_test_vectorized, y_test)
    logreg_f1score_score_last3 = f1_score(y_test, logreg_pred, average='macro', labels=[0, 1, 2])
    metrics_avg_logreg[0] += logreg_acc
    metrics_avg_logreg[1] += logreg_precision
    metrics_avg_logreg[2] += logreg_recall
    metrics_avg_logreg[3] += logreg_f1score_score_last3
    cm_logreg = confusion_matrix(y_test, logreg_pred)
    if avg_cm_logreg is None:
        avg_cm_logreg = cm_logreg
    else:
        avg_cm_logreg += cm_logreg if cm_logreg is not None else cm_logreg
    
    #SVM classifier
    svm = SVC(random_state=42)
    svm.fit(X_train_vectorized, y_train)
    svm_pred = svm.predict(X_test_vectorized)
    svm_precision, svm_recall, svm_fscore, svm_support = precision_recall_fscore_support(y_test, svm_pred, average='macro')
    svm_acc = svm.score(X_test_vectorized, y_test)
    svm_f1score_score_last3 = f1_score(y_test, svm_pred, average='macro', labels=[0, 1, 2])
    metrics_avg_svm[0] += svm_acc
    metrics_avg_svm[1] += svm_precision
    metrics_avg_svm[2] += svm_recall
    metrics_avg_svm[3] += svm_f1score_score_last3
    cm_svm = confusion_matrix(y_test, svm_pred)
    if avg_cm_svm is None:
        avg_cm_svm = cm_svm
    else:
        avg_cm_svm += cm_svm
    
    #K-NN classifier
    kNN = KNeighborsClassifier()
    kNN.fit(X_train_vectorized, y_train)
    kNN_pred = kNN.predict(X_test_vectorized)
    kNN_precision, kNN_recall, kNN_fscore, kNN_support = precision_recall_fscore_support(y_test, kNN_pred, average='macro')
    kNN_acc = svm.score(X_test_vectorized, y_test)
    kNN_f1score_score_last3 = f1_score(y_test,kNN_pred , average='macro', labels=[0, 1, 2])
    metrics_avg_kNN[0] += kNN_acc
    metrics_avg_kNN[1] += kNN_precision
    metrics_avg_kNN[2] += kNN_recall
    metrics_avg_kNN[3] += kNN_f1score_score_last3
    cm_kNN = confusion_matrix(y_test, kNN_pred)
    if avg_cm_kNN is None:
        avg_cm_kNN = cm_kNN
    else:
        avg_cm_kNN += cm_kNN
    
    if fold_count == 5:

        feature_names = vectorizer.get_feature_names()
        coefficients = logreg.coef_[0]
        coef_df = pd.DataFrame({'feature': feature_names, 'coefficient': coefficients})
        coef_df.sort_values(by='coefficient', ascending=False, inplace=True)
        print("Words with highest positive weights:")
        print(coef_df.head(20))
        print("Words with lowest negative weights:")
        print(coef_df.tail(20))
        
#All the prints        
print("Avarage Accuracy for Logistic Regression: {:.2f}%".format(100 * metrics_avg_logreg[0]/5))
print("Avarage Precision for Logistic Regression: {:.2f}%".format(100 * metrics_avg_logreg[1]/5))
print("Avarage Recall for Logistic Regression: {:.2f}%".format(100 * metrics_avg_logreg[2]/5))
print("Avarage F1-measure for Logistic Regression: {:.2f}%".format(100 * metrics_avg_logreg[3]/5))
print("The avarage confusion matrix for Logistic Regression is:")
print(avg_cm_logreg/5)

print("Avarage Accuracy for SVM: {:.2f}%".format(100 * metrics_avg_svm[0]/5))
print("Avarage Precision for SVM: {:.2f}%".format(100 * metrics_avg_svm[1]/5))
print("Avarage Recall for SVM: {:.2f}%".format(100 * metrics_avg_svm[2]/5))
print("Avarage F1-measure for SVM: {:.2f}%".format(100 * metrics_avg_svm[3]/5))
print("The avarage confusion matrix for SVM is:")
print(avg_cm_svm/5)

print("Avarage Accuracy for K-NN: {:.2f}%".format(100 * metrics_avg_kNN[0]/5))
print("Avarage Precision for K-NN: {:.2f}%".format(100 * metrics_avg_kNN[1]/5))
print("Avarage Recall for K-NN: {:.2f}%".format(100 * metrics_avg_kNN[2]/5))
print("Avarage F1-measure K-NN: SVM {:.2f}%".format(100 * metrics_avg_kNN[3]/5))
print("The avarage confusion matrix for K-NN is:")
print(avg_cm_kNN/5)

Words with highest positive weights:
           feature  coefficient
9182        burger     2.509751
23549        fries     2.159326
9206       burgers     1.214216
35483     mcdonald     1.053822
18474        drive     0.938639
8381     breakfast     0.855071
62990        wendy     0.851034
11363  cheesesteak     0.823566
63499        wings     0.764589
49677     sandwich     0.753226
40199        order     0.752353
22825         food     0.688003
11335       cheese     0.629316
11274     checkers     0.616852
34309          mac     0.616627
28751         ihop     0.605009
35484    mcdonalds     0.520513
5585         bacon     0.502462
41265     pancakes     0.479658
63459       window     0.474651
Words with lowest negative weights:
          feature  coefficient
17471      dishes    -0.470386
53611        soup    -0.473634
35614   meatballs    -0.478196
28492         hut    -0.493847
30666    japanese    -0.497871
9466         byob    -0.502768
16266   delicious    -0.514812
17204  

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


Based on the features with highest positive weights and the features with lowest negative weights The features with the highest positive weights suggest that the classifier is associating these features with the "Burgers" category, as all of these words are related to fast food and American cuisine.On the other hand, the features with the lowest negative weights suggest that the classifier is associating these features with the "Japanese" and "Italian" categories.From this, we can conclude that the classifier is able to differentiate between the three categories to some degree, but there is room for improvement in terms of reducing the overlap between the "Burgers" category  with the  "Japanese" or "Italian" categories.



                               -Step 2-

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix

import gensim 
import gensim.models
from gensim.models import Word2Vec
from gensim import utils
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from gensim.models import Doc2Vec
from gensim.models.doc2vec import TaggedDocument

word_by_word_X = [utils.simple_preprocess(sentence) for sentence in X]

In [17]:
kfold = KFold(n_splits=5, shuffle=True, random_state=42)

cbow_metrics_avg_logreg = np.zeros(4)
cbow_metrics_avg_svm = np.zeros(4)
cbow_metrics_avg_kNN = np.zeros(4)
cbow_avg_cm_logreg = None
cbow_avg_cm_svm = None
cbow_avg_cm_kNN= None

doc2vec_metrics_avg_logreg = np.zeros(4)
doc2vec_metrics_avg_svm = np.zeros(4)
doc2vec_metrics_avg_kNN = np.zeros(4)
doc2vec_avg_cm_logreg = None
doc2vec_avg_cm_svm = None
doc2vec_avg_cm_kNN= None

skip_gram_metrics_avg_logreg = np.zeros(4)
skip_gram_metrics_avg_svm = np.zeros(4)
skip_gram_metrics_avg_kNN = np.zeros(4)
skip_gram_avg_cm_logreg = None
skip_gram_avg_cm_svm = None
skip_gram_avg_cm_kNN= None

# Split the data into 5 folds
for train_index, test_index in kfold.split(word_by_word_X):
    X_train, X_test = [word_by_word_X[i] for i in train_index], [word_by_word_X[i] for i in test_index]
    y_train, y_test = [y[i] for i in train_index], [y[i] for i in test_index]
   
    #Chow Model 
    cbow_model = Word2Vec(sentences=X_train, vector_size=50, window=10, min_count=1, workers=4)    
    cbow_model.train(X_train, total_examples=len(X_train), epochs=10)
    X_avg_train_cbow = [np.array([cbow_model.wv[x] for x in y]).mean(axis = 0) for y in X_train]
    X_avg_test_cbow = [np.array([cbow_model.wv[x] for x in y if x in cbow_model.wv]).mean(axis = 0) for y in X_test]
    #SkipGram Model 
    skip_gram_model = Word2Vec(sentences=X_train, vector_size=50, window=10, min_count=1, workers=4, sg=1)
    skip_gram_model.train(X_train, total_examples=len(X_train), epochs=10)
    X_avg_train_skip_gram = [np.array([skip_gram_model.wv[x] for x in y]).mean(axis = 0) for y in X_train]
    X_avg_test_skip_gram = [np.array([skip_gram_model.wv[x] for x in y if x in skip_gram_model.wv]).mean(axis = 0) for y in X_test]

    
    
    #Doc2Vec model
    tagged_X_train = [gensim.models.doc2vec.TaggedDocument(X_train[i], [i]) for i in range(len(X_train))]
    d2v_model = gensim.models.doc2vec.Doc2Vec(vector_size=50, min_count=2, epochs=40)
    d2v_model.build_vocab(tagged_X_train)
    d2v_model.train(tagged_X_train, total_examples=d2v_model.corpus_count, epochs=d2v_model.epochs)
    X_train_d2v = [d2v_model.infer_vector(x) for x in X_train]
    X_test_d2v = [d2v_model.infer_vector(x) for x in X_test]
    
    
    
    #classifiers for chow model
    lr_clf = LogisticRegression(random_state=0, solver='lbfgs', multi_class='auto', n_jobs=-1)
    lr_clf.fit(np.array(X_avg_train_cbow), np.array(y_train))
    lr_clf_pred = lr_clf.predict(X_avg_test_cbow)
    lr_clf_precision = precision_score(y_test, lr_clf_pred, average='macro') 
    lr_clf_f1 = f1_score(y_test, lr_clf_pred, average='macro', labels=[0, 1, 2])
    lr_clf_accuracy = accuracy_score(y_test, lr_clf_pred)
    lr_clf_recall = recall_score(y_test, lr_clf_pred, average='macro')

    cbow_metrics_avg_logreg[0] += lr_clf_accuracy
    cbow_metrics_avg_logreg[1] += lr_clf_precision
    cbow_metrics_avg_logreg[2] += lr_clf_recall
    cbow_metrics_avg_logreg[3] += lr_clf_f1
    cbow_cm_logreg = confusion_matrix(y_test, lr_clf_pred)
    if cbow_avg_cm_logreg is None:
        cbow_avg_cm_logreg = cbow_cm_logreg
    else:
        cbow_avg_cm_logreg += cbow_cm_logreg 
        
    svm = SVC(random_state=42)
    svm.fit(X_avg_train_cbow, y_train)
    svm_pred = svm.predict(X_avg_test_cbow)
    svm_precision = precision_score(y_test, svm_pred, average='macro') 
    svm_f1 = f1_score(y_test, svm_pred, average='macro', labels=[0, 1, 2])
    svm_accuracy = accuracy_score(y_test, svm_pred)
    svm_recall = recall_score(y_test, svm_pred, average='macro')

    cbow_metrics_avg_svm[0] += svm_accuracy
    cbow_metrics_avg_svm[1] += svm_precision
    cbow_metrics_avg_svm[2] += svm_recall
    cbow_metrics_avg_svm[3] += svm_f1
    cbow_cm_svm = confusion_matrix(y_test, svm_pred)
    if cbow_avg_cm_svm is None:
        cbow_avg_cm_svm = cbow_cm_svm
    else:
        cbow_avg_cm_svm += cbow_cm_svm
    

    kNN = KNeighborsClassifier()
    kNN.fit(np.array(X_avg_train_cbow), np.array(y_train))
    kNN_pred = lr_clf.predict(X_avg_test_cbow)
    kNN_precision = precision_score(y_test, kNN_pred, average='macro') 
    kNN_f1 = f1_score(y_test, kNN_pred, average='macro', labels=[0, 1, 2])
    kNN_accuracy = accuracy_score(y_test, kNN_pred)
    kNN_recall = recall_score(y_test, kNN_pred, average='macro')

    cbow_metrics_avg_kNN[0] += kNN_accuracy
    cbow_metrics_avg_kNN[1] += kNN_precision
    cbow_metrics_avg_kNN[2] += kNN_recall
    cbow_metrics_avg_kNN[3] += kNN_f1
    cbow_cm_kNN = confusion_matrix(y_test, kNN_pred)
    if cbow_avg_cm_kNN is None:
        cbow_avg_cm_kNN = cbow_cm_kNN
    else:
        cbow_avg_cm_kNN += cbow_cm_kNN

        
    #classifiers for skipgram model
    lr_clf = LogisticRegression(random_state=0, solver='lbfgs', multi_class='auto', n_jobs=-1)
    lr_clf.fit(np.array(X_avg_train_skip_gram), np.array(y_train))
    lr_clf_pred = lr_clf.predict(X_avg_test_skip_gram)
    lr_clf_precision = precision_score(y_test, lr_clf_pred, average='macro') 
    lr_clf_f1 = f1_score(y_test, lr_clf_pred, average='macro', labels=[0, 1, 2])
    lr_clf_accuracy = accuracy_score(y_test, lr_clf_pred)
    lr_clf_recall = recall_score(y_test, lr_clf_pred, average='macro')

    skip_gram_metrics_avg_logreg[0] += lr_clf_accuracy
    skip_gram_metrics_avg_logreg[1] += lr_clf_precision
    skip_gram_metrics_avg_logreg[2] += lr_clf_recall
    skip_gram_metrics_avg_logreg[3] += lr_clf_f1
    skip_gram_cm_logreg = confusion_matrix(y_test, lr_clf_pred)
    if skip_gram_avg_cm_logreg is None:
        skip_gram_avg_cm_logreg = skip_gram_cm_logreg
    else:
        skip_gram_avg_cm_logreg += skip_gram_cm_logreg 
        
    svm = SVC(random_state=42)
    svm.fit(X_avg_train_skip_gram, y_train)
    svm_pred = svm.predict(X_avg_test_skip_gram)
    svm_precision = precision_score(y_test, svm_pred, average='macro') 
    svm_f1 = f1_score(y_test, svm_pred, average='macro', labels=[0, 1, 2])
    svm_accuracy = accuracy_score(y_test, svm_pred)
    svm_recall = recall_score(y_test, svm_pred, average='macro')

    skip_gram_metrics_avg_svm[0] += svm_accuracy
    skip_gram_metrics_avg_svm[1] += svm_precision
    skip_gram_metrics_avg_svm[2] += svm_recall
    skip_gram_metrics_avg_svm[3] += svm_f1
    skip_gram_cm_svm = confusion_matrix(y_test, svm_pred)
    if skip_gram_avg_cm_svm is None:
        skip_gram_avg_cm_svm = skip_gram_cm_svm
    else:
        skip_gram_avg_cm_svm += skip_gram_cm_svm
    

    kNN = KNeighborsClassifier()
    kNN.fit(np.array(X_avg_train_skip_gram), np.array(y_train))
    kNN_pred = lr_clf.predict(X_avg_test_skip_gram)
    kNN_precision = precision_score(y_test, kNN_pred, average='macro') 
    kNN_f1 = f1_score(y_test, kNN_pred, average='macro', labels=[0, 1, 2])
    kNN_accuracy = accuracy_score(y_test, kNN_pred)
    kNN_recall = recall_score(y_test, kNN_pred, average='macro')

    skip_gram_metrics_avg_kNN[0] += kNN_accuracy
    skip_gram_metrics_avg_kNN[1] += kNN_precision
    skip_gram_metrics_avg_kNN[2] += kNN_recall
    skip_gram_metrics_avg_kNN[3] += kNN_f1
    skip_gram_cm_kNN = confusion_matrix(y_test, kNN_pred)
    if skip_gram_avg_cm_kNN is None:
        skip_gram_avg_cm_kNN = skip_gram_cm_kNN
    else:
        skip_gram_avg_cm_kNN += skip_gram_cm_kNN

        
    
    
    #classifiers for Doc2Vec model
    lr_clf = LogisticRegression(random_state=0, solver='lbfgs', multi_class='auto', n_jobs=-1)
    lr_clf.fit(np.array(X_train_d2v), np.array(y_train))
    lr_clf_pred = lr_clf.predict(X_test_d2v)
    lr_clf_precision = precision_score(y_test, lr_clf_pred, average='macro') 
    lr_clf_f1 = f1_score(y_test, lr_clf_pred, average='macro', labels=[0, 1, 2])
    lr_clf_accuracy = accuracy_score(y_test, lr_clf_pred)
    lr_clf_recall = recall_score(y_test, lr_clf_pred, average='macro')

    doc2vec_metrics_avg_logreg[0] += lr_clf_accuracy
    doc2vec_metrics_avg_logreg[1] += lr_clf_precision
    doc2vec_metrics_avg_logreg[2] += lr_clf_recall
    doc2vec_metrics_avg_logreg[3] += lr_clf_f1
    doc2vec_cm_logreg = confusion_matrix(y_test, lr_clf_pred)
    if doc2vec_avg_cm_logreg is None:
        doc2vec_avg_cm_logreg = doc2vec_cm_logreg
    else:
        doc2vec_avg_cm_logreg += doc2vec_cm_logreg 
        
    svm = SVC(random_state=42)
    svm.fit(X_train_d2v, y_train)
    svm_pred = svm.predict(X_test_d2v)
    svm_precision = precision_score(y_test, svm_pred, average='macro') 
    svm_f1 = f1_score(y_test, svm_pred, average='macro', labels=[0, 1, 2])
    svm_accuracy = accuracy_score(y_test, svm_pred)
    svm_recall = recall_score(y_test, svm_pred, average='macro')

    doc2vec_metrics_avg_svm[0] += svm_accuracy
    doc2vec_metrics_avg_svm[1] += svm_precision
    doc2vec_metrics_avg_svm[2] += svm_recall
    doc2vec_metrics_avg_svm[3] += svm_f1
    doc2vec_cm_svm = confusion_matrix(y_test, svm_pred)
    if doc2vec_avg_cm_svm is None:
        doc2vec_avg_cm_svm = doc2vec_cm_svm
    else:
        doc2vec_avg_cm_svm += doc2vec_cm_svm
    

    kNN = KNeighborsClassifier()
    kNN.fit(np.array(X_train_d2v), np.array(y_train))
    kNN_pred = lr_clf.predict(X_test_d2v)
    kNN_precision = precision_score(y_test, kNN_pred, average='macro') 
    kNN_f1 = f1_score(y_test, kNN_pred, average='macro', labels=[0, 1, 2])
    kNN_accuracy = accuracy_score(y_test, kNN_pred)
    kNN_recall = recall_score(y_test, kNN_pred, average='macro')

    doc2vec_metrics_avg_kNN[0] += kNN_accuracy
    doc2vec_metrics_avg_kNN[1] += kNN_precision
    doc2vec_metrics_avg_kNN[2] += kNN_recall
    doc2vec_metrics_avg_kNN[3] += kNN_f1
    doc2vec_cm_kNN = confusion_matrix(y_test, kNN_pred)
    if doc2vec_avg_cm_kNN is None:
        doc2vec_avg_cm_kNN = doc2vec_cm_kNN
    else:
        doc2vec_avg_cm_kNN += doc2vec_cm_kNN

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-le

In [18]:
print("For CBOW model: ")
print("Avarage Accuracy for Logistic Regression: {:.2f}%".format(100 * cbow_metrics_avg_logreg[0]/5))
print("Avarage Precision for Logistic Regression: {:.2f}%".format(100 * cbow_metrics_avg_logreg[1]/5))
print("Avarage Recall for Logistic Regression: {:.2f}%".format(100 * cbow_metrics_avg_logreg[2]/5))
print("Avarage F1-measure for Logistic Regression: {:.2f}%".format(100 * cbow_metrics_avg_logreg[3]/5))
print("The avarage confusion matrix for Logistic Regression is:")
print(cbow_avg_cm_logreg/5)

print("Avarage Accuracy for SVM: {:.2f}%".format(100 * cbow_metrics_avg_svm[0]/5))
print("Avarage Precision for SVM: {:.2f}%".format(100 * cbow_metrics_avg_svm[1]/5))
print("Avarage Recall for SVM: {:.2f}%".format(100 * cbow_metrics_avg_svm[2]/5))
print("Avarage F1-measure for SVM: {:.2f}%".format(100 * cbow_metrics_avg_svm[3]/5))
print("The avarage confusion matrix for SVM is:")
print(cbow_avg_cm_svm/5)

print("Avarage Accuracy for K-NN: {:.2f}%".format(100 * cbow_metrics_avg_kNN[0]/5))
print("Avarage Precision for K-NN: {:.2f}%".format(100 * cbow_metrics_avg_kNN[1]/5))
print("Avarage Recall for K-NN: {:.2f}%".format(100 * cbow_metrics_avg_kNN[2]/5))
print("Avarage F1-measure K-NN: SVM {:.2f}%".format(100 * cbow_metrics_avg_kNN[3]/5))
print("The avarage confusion matrix for K-NN is:")
print(cbow_avg_cm_kNN/5)

For CBOW model: 
Avarage Accuracy for Logistic Regression: 91.80%
Avarage Precision for Logistic Regression: 92.72%
Avarage Recall for Logistic Regression: 92.40%
Avarage F1-measure for Logistic Regression: 92.52%
The avarage confusion matrix for Logistic Regression is:
[[56.2  6.6  0.4]
 [ 7.  78.8  0.4]
 [ 0.2  1.  39.6]]
Avarage Accuracy for SVM: 91.38%
Avarage Precision for SVM: 92.33%
Avarage Recall for SVM: 91.65%
Avarage F1-measure for SVM: 91.95%
The avarage confusion matrix for SVM is:
[[55.4  7.2  0.6]
 [ 6.2 79.6  0.4]
 [ 1.   1.  38.8]]
Avarage Accuracy for K-NN: 91.80%
Avarage Precision for K-NN: 92.72%
Avarage Recall for K-NN: 92.40%
Avarage F1-measure K-NN: SVM 92.52%
The avarage confusion matrix for K-NN is:
[[56.2  6.6  0.4]
 [ 7.  78.8  0.4]
 [ 0.2  1.  39.6]]


In [19]:
print("For Skipgram model: ")
print("Avarage Accuracy for Logistic Regression: {:.2f}%".format(100 * skip_gram_metrics_avg_logreg[0]/5))
print("Avarage Precision for Logistic Regression: {:.2f}%".format(100 * skip_gram_metrics_avg_logreg[1]/5))
print("Avarage Recall for Logistic Regression: {:.2f}%".format(100 * skip_gram_metrics_avg_logreg[2]/5))
print("Avarage F1-measure for Logistic Regression: {:.2f}%".format(100 * skip_gram_metrics_avg_logreg[3]/5))
print("The avarage confusion matrix for Logistic Regression is:")
print(skip_gram_avg_cm_logreg/5)

print("Avarage Accuracy for SVM: {:.2f}%".format(100 * skip_gram_metrics_avg_svm[0]/5))
print("Avarage Precision for SVM: {:.2f}%".format(100 * skip_gram_metrics_avg_svm[1]/5))
print("Avarage Recall for SVM: {:.2f}%".format(100 * skip_gram_metrics_avg_svm[2]/5))
print("Avarage F1-measure for SVM: {:.2f}%".format(100 * skip_gram_metrics_avg_svm[3]/5))
print("The avarage confusion matrix for SVM is:")
print(skip_gram_avg_cm_svm/5)

print("Avarage Accuracy for K-NN: {:.2f}%".format(100 * skip_gram_metrics_avg_kNN[0]/5))
print("Avarage Precision for K-NN: {:.2f}%".format(100 * skip_gram_metrics_avg_kNN[1]/5))
print("Avarage Recall for K-NN: {:.2f}%".format(100 * skip_gram_metrics_avg_kNN[2]/5))
print("Avarage F1-measure K-NN: SVM {:.2f}%".format(100 * skip_gram_metrics_avg_kNN[3]/5))
print("The avarage confusion matrix for K-NN is:")
print(skip_gram_avg_cm_kNN/5)

For Skipgram model: 
Avarage Accuracy for Logistic Regression: 86.33%
Avarage Precision for Logistic Regression: 88.92%
Avarage Recall for Logistic Regression: 86.01%
Avarage F1-measure for Logistic Regression: 87.16%
The avarage confusion matrix for Logistic Regression is:
[[48.6 14.4  0.2]
 [ 6.6 79.2  0.4]
 [ 0.6  3.8 36.4]]
Avarage Accuracy for SVM: 89.59%
Avarage Precision for SVM: 91.78%
Avarage Recall for SVM: 88.86%
Avarage F1-measure for SVM: 90.00%
The avarage confusion matrix for SVM is:
[[51.  12.   0.2]
 [ 3.2 82.6  0.4]
 [ 1.   3.  36.8]]
Avarage Accuracy for K-NN: 86.33%
Avarage Precision for K-NN: 88.92%
Avarage Recall for K-NN: 86.01%
Avarage F1-measure K-NN: SVM 87.16%
The avarage confusion matrix for K-NN is:
[[48.6 14.4  0.2]
 [ 6.6 79.2  0.4]
 [ 0.6  3.8 36.4]]


In [20]:
print("For Doc2Vec model: ")
print("Avarage Accuracy for Logistic Regression: {:.2f}%".format(100 * doc2vec_metrics_avg_logreg[0]/5))
print("Avarage Precision for Logistic Regression: {:.2f}%".format(100 * doc2vec_metrics_avg_logreg[1]/5))
print("Avarage Recall for Logistic Regression: {:.2f}%".format(100 * doc2vec_metrics_avg_logreg[2]/5))
print("Avarage F1-measure for Logistic Regression: {:.2f}%".format(100 * doc2vec_metrics_avg_logreg[3]/5))
print("The avarage confusion matrix for Logistic Regression is:")
print(doc2vec_avg_cm_logreg/5)

print("Avarage Accuracy for SVM: {:.2f}%".format(100 * doc2vec_metrics_avg_svm[0]/5))
print("Avarage Precision for SVM: {:.2f}%".format(100 * doc2vec_metrics_avg_svm[1]/5))
print("Avarage Recall for SVM: {:.2f}%".format(100 * doc2vec_metrics_avg_svm[2]/5))
print("Avarage F1-measure for SVM: {:.2f}%".format(100 * doc2vec_metrics_avg_svm[3]/5))
print("The avarage confusion matrix for SVM is:")
print(doc2vec_avg_cm_svm/5)

print("Avarage Accuracy for K-NN: {:.2f}%".format(100 * doc2vec_metrics_avg_kNN[0]/5))
print("Avarage Precision for K-NN: {:.2f}%".format(100 * doc2vec_metrics_avg_kNN[1]/5))
print("Avarage Recall for K-NN: {:.2f}%".format(100 * doc2vec_metrics_avg_kNN[2]/5))
print("Avarage F1-measure K-NN: SVM {:.2f}%".format(100 * doc2vec_metrics_avg_kNN[3]/5))
print("The avarage confusion matrix for K-NN is:")
print(doc2vec_avg_cm_kNN/5)

For Doc2Vec model: 
Avarage Accuracy for Logistic Regression: 90.33%
Avarage Precision for Logistic Regression: 91.17%
Avarage Recall for Logistic Regression: 90.83%
Avarage F1-measure for Logistic Regression: 90.96%
The avarage confusion matrix for Logistic Regression is:
[[54.8  7.6  0.8]
 [ 7.6 78.2  0.4]
 [ 1.2  0.8 38.8]]
Avarage Accuracy for SVM: 92.74%
Avarage Precision for SVM: 93.60%
Avarage Recall for SVM: 93.27%
Avarage F1-measure for SVM: 93.39%
The avarage confusion matrix for SVM is:
[[55.6  7.2  0.4]
 [ 5.2 80.6  0.4]
 [ 0.2  0.4 40.2]]
Avarage Accuracy for K-NN: 90.33%
Avarage Precision for K-NN: 91.17%
Avarage Recall for K-NN: 90.83%
Avarage F1-measure K-NN: SVM 90.96%
The avarage confusion matrix for K-NN is:
[[54.8  7.6  0.8]
 [ 7.6 78.2  0.4]
 [ 1.2  0.8 38.8]]


The results show that the TF-IDF model performed the best, followed by CBOW and Doc2Vec, with the Skipgram model performing the worst.The evaluation metrics used for all three models (tf-idf, Skipgram, and Doc2Vec) were accuracy, precision, recall, and F1-score. The average accuracy for all models was above 86%, which indicates that the models are performing well in terms of predicting the correct class.
    For the TF-IDF approach, the average accuracy for SVM was 92.11%, with an average precision of 93.11%, recall of 92.38%, and F1-measure of 92.71%. The confusion matrix was similar to that of logistic regression, with some confusion between classes 1 and 2.
    For the CBOW model, all three classifiers performed similarly as well, with average accuracy around 91%. The average precision and recall were also around 92%, and the average F1-measure was around 92.5%. The confusion matrix was similar to that of the TF-IDF model.
    For the Doc2Vec model, the SVM classifier performed the best, with an average accuracy of 92.74%, while the Logistic Regression classifier performed slightly worse, with an average accuracy of 90.33%. The average precision, recall, and F1-measure were all around 91-93%. The confusion matrix was similar to that of the TF-IDF and CBOW models.
    For the Skipgram model, the SVM classifier performed the best, with an average accuracy of 89.59%, followed by the Logistic Regression classifier, with an average accuracy of 86.33%. The average precision, recall, and F1-measure were all around 87-91%. The confusion matrix suggests that there may be some overlap or ambiguity between the classes
   Overall SVM perfomed the best for all the above models.

                                             -Bonus-

In [10]:
import gensim.downloader as api
path = api.load("word2vec-google-news-300", return_path=True)
print(path)

/Users/billyvog/gensim-data/word2vec-google-news-300/word2vec-google-news-300.gz


In [11]:
g_model = gensim.models.KeyedVectors.load_word2vec_format(path, binary=True)  

In [13]:
g_model_metrics_avg_logreg = np.zeros(4)
g_model_avg_cm_logreg = None


for train_index, test_index in kfold.split(word_by_word_X):
    X_train, X_test = [word_by_word_X[i] for i in train_index], [word_by_word_X[i] for i in test_index]
    y_train, y_test = [y[i] for i in train_index], [y[i] for i in test_index]
    
    train_gmodel = [[g_model[x] for x in y if x in g_model] for y in X_train]
    train_data_labels = [(x,y) for (x,y) in zip(train_gmodel, y_train) if len(x) > 0]
    X_train_gm = [np.array(x) for (x,y) in train_data_labels]
    y_train_gm = [y for (x,y) in train_data_labels]
    X_train_gmodel = [x.mean(axis = 0) for x in X_train_gm]
    test_gmodel = [[g_model[x] for x in y if x in g_model] for y in X_test]
    test_data_labels = [(x,y) for (x,y) in zip(test_gmodel, y_test) if len(x) > 0]
    X_test_gm = [np.array(x) for (x,y) in test_data_labels]
    y_test_gm = [y for (x,y) in test_data_labels]

    X_test_gmodel = [x.mean(axis = 0) for x in X_test_gm]
    #classifiers for Doc2Vec model
    lr_clf = LogisticRegression(random_state=0, solver='lbfgs', multi_class='auto', n_jobs=-1)
    lr_clf.fit(np.array(X_train_gmodel), np.array(y_train_gm))
    lr_clf_pred = lr_clf.predict(test_gmodel)
    lr_clf_precision = precision_score(y_test, lr_clf_pred, average='macro') 
    lr_clf_f1 = f1_score(y_test_gm, lr_clf_pred, average='macro', labels=[0, 1, 2])
    lr_clf_accuracy = accuracy_score(y_test_gm, lr_clf_pred)
    lr_clf_recall = recall_score(y_test_gm, lr_clf_pred, average='macro')

    g_model_metrics_avg_logreg[0] += lr_clf_accuracy
    g_model_metrics_avg_logreg[1] += lr_clf_precision
    g_model_metrics_avg_logreg[2] += lr_clf_recall
    g_model_metrics_avg_logreg[3] += lr_clf_f1
    g_model_cm_logreg = confusion_matrix(y_test, lr_clf_pred)
    if g_model_avg_cm_logreg is None:
        g_model_avg_cm_logreg = g_model_cm_logreg
    else:
        g_model_avg_cm_logreg += g_model_cm_logreg 
        
    

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/utils/validation.py:746: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  array = np.asarray(array, order=order, dtype=dtype)


KeyboardInterrupt: 

In [ ]:
print("For Google word2vec model: ")
print("Avarage Accuracy for Logistic Regression: {:.2f}%".format(100 * g_model_metrics_avg_logreg[0]/5))
print("Avarage Precision for Logistic Regression: {:.2f}%".format(100 * g_model_metrics_avg_logreg[1]/5))
print("Avarage Recall for Logistic Regression: {:.2f}%".format(100 * g_model_metrics_avg_logreg[2]/5))
print("Avarage F1-measure for Logistic Regression: {:.2f}%".format(100 * g_model_metrics_avg_logreg[3]/5))
print("The avarage confusion matrix for Logistic Regression is:")
print(g_model_avg_cm_logreg/5)

took forever so i couldnt test it